

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Installation &mdash; ACTIONet  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="ACTIONet" href="ACTIONet/modules.html" />
    <link rel="prev" title="Welcome to ACTIONet’s documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> ACTIONet
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="ACTIONet/modules.html">ACTIONet</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorials/modules.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ACTIONet</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Installation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/INTRODUCTION.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>### Setting Up the Environment (Preinstallation)
<strong>For Linux Users</strong></p>
<p>Verify that the cmake version you are using is &gt;=3.19.</p>
<p>For the optimal performance on Intel-based architectures, installing [Intel Math Kernel Library (MKL)](<a class="reference external" href="https://software.intel.com/content/www/us/en/develop/articles/intel-math-kernel-library-intel-mkl-2020-install-guide.html">https://software.intel.com/content/www/us/en/develop/articles/intel-math-kernel-library-intel-mkl-2020-install-guide.html</a>) is <strong>highly</strong> recommended. After installing, make sure <cite>MKLROOT</cite> is defined by running the [setvars](<a class="reference external" href="https://software.intel.com/content/www/us/en/develop/documentation/using-configuration-file-for-setvars-sh/top.html">https://software.intel.com/content/www/us/en/develop/documentation/using-configuration-file-for-setvars-sh/top.html</a>) script.</p>
<p><strong>Install library dependencies</strong>
To install the <cite>ACTIONet</cite> dependencie on debian-based linux machines, run:</p>
<p><code class="docutils literal notranslate"><span class="pre">`bash</span>
<span class="pre">sudo</span> <span class="pre">apt-get</span> <span class="pre">install</span> <span class="pre">libhdf5-dev</span> <span class="pre">libsuitesparse-dev</span> <span class="pre">libnss3</span> <span class="pre">xvfb</span>
<span class="pre">`</span></code></p>
<p>For Mac-based systems, you can use [brew](<a class="reference external" href="https://brew.sh/">https://brew.sh/</a>) instead:</p>
<p><code class="docutils literal notranslate"><span class="pre">`bash</span>
<span class="pre">brew</span> <span class="pre">install</span> <span class="pre">hdf5</span> <span class="pre">suite-sparse</span> <span class="pre">c-blosc</span>
<span class="pre">`</span></code></p>
<p>### Installing ACTIONet Python Package
Use <cite>pip</cite> to install ACTIONet directly from this repository:
<code class="docutils literal notranslate"><span class="pre">`bash</span>
<span class="pre">pip</span> <span class="pre">install</span> <span class="pre">git+https://github.com/shmohammadi86/ACTIONet&#64;python-devel</span>
<span class="pre">`</span></code></p>
<p>To install from source:
<code class="docutils literal notranslate"><span class="pre">`</span>
<span class="pre">git</span> <span class="pre">clone</span> <span class="pre">--recurse-submodules</span> <span class="pre">https://github.com/shmohammadi86/ACTIONet.git</span>
<span class="pre">git</span> <span class="pre">submodule</span> <span class="pre">update</span> <span class="pre">--init</span>
<span class="pre">python</span> <span class="pre">setup.py</span> <span class="pre">build</span>
<span class="pre">python</span> <span class="pre">setup.py</span> <span class="pre">develop</span>
<span class="pre">`</span></code></p>
<p># Running ACTIONet
<strong>Note</strong> If you are using <cite>MKL</cite>, make sure to properly [set the number of threads](<a class="reference external" href="https://software.intel.com/content/www/us/en/develop/documentation/mkl-macos-developer-guide/top/managing-performance-and-memory/improving-performance-with-threading/techniques-to-set-the-number-of-threads.html">https://software.intel.com/content/www/us/en/develop/documentation/mkl-macos-developer-guide/top/managing-performance-and-memory/improving-performance-with-threading/techniques-to-set-the-number-of-threads.html</a>) used prior to running <cite>ACTIONet</cite>.</p>
<p>## Example Run
Here is a simple example to get you started:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">urllib.request</span>

<span class="kn">import</span> <span class="nn">ACTIONet</span> <span class="k">as</span> <span class="nn">an</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

<span class="c1"># Download example dataset from the 10X Genomics website</span>
<span class="n">opener</span> <span class="o">=</span> <span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">build_opener</span><span class="p">()</span>
<span class="n">opener</span><span class="o">.</span><span class="n">addheaders</span> <span class="o">=</span> <span class="p">[(</span><span class="s1">&#39;User-agent&#39;</span><span class="p">,</span> <span class="s1">&#39;Mozilla/5.0&#39;</span><span class="p">)]</span>
<span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">install_opener</span><span class="p">(</span><span class="n">opener</span><span class="p">)</span>
<span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlretrieve</span><span class="p">(</span><span class="s1">&#39;http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_10k_v3/pbmc_10k_v3_filtered_feature_bc_matrix.h5&#39;</span><span class="p">,</span> <span class="s1">&#39;pbmc_10k_v3.h5&#39;</span><span class="p">)</span>

<span class="c1"># Read and filter the data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_h5</span><span class="p">(</span><span class="s1">&#39;pbmc_10k_v3.h5&#39;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">(</span><span class="n">join</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="n">an</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells_per_feature</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">min_features_per_cell</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="c1"># Run ACTIONet</span>
<span class="n">an</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">reduce_kernel</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">an</span><span class="o">.</span><span class="n">run_ACTIONet</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="c1"># Annotate cell-types</span>
<span class="n">marker_genes</span><span class="p">,</span> <span class="n">directions</span><span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="n">an</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">load_markers</span><span class="p">(</span><span class="s1">&#39;PBMC_Monaco2019_12celltypes&#39;</span><span class="p">)</span>
<span class="n">cell_labels</span><span class="p">,</span> <span class="n">confidences</span><span class="p">,</span> <span class="n">Z</span> <span class="o">=</span> <span class="n">an</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">annotate_cells_using_markers</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="n">directions</span><span class="p">,</span> <span class="n">names</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltypes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_labels</span>

<span class="c1"># Visualize output</span>
<span class="n">an</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">plot_ACTIONet</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;celltypes&#39;</span><span class="p">,</span> <span class="n">transparency_key</span><span class="o">=</span><span class="s1">&#39;node_centrality&#39;</span><span class="p">)</span>

<span class="c1"># Export results</span>
<span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;pbmc_10k_v3.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>## Visualizing results using cellxgene</p>
<p>The output of ACTIONet in the python implementation is internally stored as as <cite>AnnData</cite> object, and R <cite>ACE</cite> objects can be imported from/exported to <cite>AnnData</cite> using functions <cite>AnnData2ACE()</cite> and <cite>ACE2AnnData()</cite> functions, respectively. <cite>AnnData</cite> objects can be directly loaded into [cellxgene](<a class="reference external" href="https://github.com/chanzuckerberg/cellxgene">https://github.com/chanzuckerberg/cellxgene</a>) package, an open-source viewer for interactive single-cell data visualization. <cite>cellxgene</cite> can be installed as:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install cellxgene
</pre></div>
</div>
<p>Then to visualize the results of ACTIONet, run:
.. code-block:: bash</p>
<blockquote>
<div><p>cellxgene launch pbmc_10k_v3.h5ad</p>
</div></blockquote>
<p>where <em>pbmc_10k_v3.h5ad</em> is the name of the file we exported using <cite>adata.write()</cite> function.</p>
<p># Additional tutorials
You can access ACTIONet tutorials from:
1. [ACTIONet framework at a glance (human PBMC 3k dataset)](<a class="reference external" href="http://compbio.mit.edu/ACTIONet/tutorials/mini_intro.html">http://compbio.mit.edu/ACTIONet/tutorials/mini_intro.html</a>)
2. [Introduction to the ACTIONet framework (human PBMC Granja et al. dataset)](<a class="reference external" href="http://compbio.mit.edu/ACTIONet/tutorials/intro.html">http://compbio.mit.edu/ACTIONet/tutorials/intro.html</a>)
3. [Introduction to cluster-centric analysis using the ACTIONet framework](<a class="reference external" href="http://compbio.mit.edu/ACTIONet/tutorials/clustering.html">http://compbio.mit.edu/ACTIONet/tutorials/clustering.html</a>)
4. [To batch correct or not to batch correct, that is the question!](<a class="reference external" href="http://compbio.mit.edu/ACTIONet/tutorials/batch.html">http://compbio.mit.edu/ACTIONet/tutorials/batch.html</a>)
5. [PortingData: Import/export options in the ACTIONet framework](<a class="reference external" href="http://compbio.mit.edu/ACTIONet/tutorials/porting_data.html">http://compbio.mit.edu/ACTIONet/tutorials/porting_data.html</a>)
6. [Interactive visualization, annotation, and exploration](<a class="reference external" href="http://compbio.mit.edu/ACTIONet/tutorials/annotation.html">http://compbio.mit.edu/ACTIONet/tutorials/annotation.html</a>)
7. [Constructing cell-type/cell-state-specific networks using SCINET](<a class="reference external" href="http://compbio.mit.edu/ACTIONet/tutorials/scinet.html">http://compbio.mit.edu/ACTIONet/tutorials/scinet.html</a>)</p>
<p>You can also find a [step-by-step guide](<a class="reference external" href="http://compbio.mit.edu/ACTIONet/tutorials/guide.html">http://compbio.mit.edu/ACTIONet/tutorials/guide.html</a>) to learning the core functionalities of the ACTIONet framework.</p>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="ACTIONet/modules.html" class="btn btn-neutral float-right" title="ACTIONet" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to ACTIONet’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Shahin Mohammadi.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>